Build failed of crc32c

### Company or project name

_No response_

### Describe the problem

Build failed with contrib/crc32c.If we change the command line `-march=native` with `-march=armv8.1a+aes`, build succeed.

### Operating system

openEuler 24.03-LTS-SP2

### CMake version

cmake version 4.1.1

### Ninja version

1.13.1

### Compiler name and version

clang version 20.1.8 (https://github.com/llvm/llvm-project 87f0227cb60147a26a1eeb4fb06e3b509e9c7261)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin

### Full cmake and/or ninja output with the error

```
FAILED: [code=1] contrib/crc32c-cmake/CMakeFiles/_crc32c.dir/__/crc32c/src/crc32c_arm64.cc.o 
/usr/bin/ccache /usr/local/bin/clang++ --target=aarch64-linux-gnu --sysroot=/root/ClickHouse/cmake/linux/../../contrib/sysroot/linux-aarch64/aarch64-linux-gnu/libc -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -D_LIBUNWIND_IS_NATIVE_ONLY -I/root/ck-build/include -I/root/ClickHouse/contrib/crc32c-cmake/aarch64 -I/root/ClickHouse/contrib/crc32c/include -I/root/ck-build/contrib/llvm-project/libcxx/include/c++/v1 -isystem /root/ClickHouse/contrib/llvm-project/libcxx/include -isystem /root/ClickHouse/contrib/llvm-project/libcxxabi/include -isystem /root/ClickHouse/contrib/llvm-project/libunwind/include --gcc-toolchain=/root/ClickHouse/cmake/linux/../../contrib/sysroot/linux-aarch64 -fdiagnostics-color=always -fsized-deallocation  -pipe -march=native -fasynchronous-unwind-tables -ffile-prefix-map=/root/ClickHouse=. -falign-functions=64 -ffp-contract=off  -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fPIC -w -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++23   -D OS_LINUX -Wno-deprecated-declarations -Wno-poison-system-directories -nostdinc++ -MD -MT contrib/crc32c-cmake/CMakeFiles/_crc32c.dir/__/crc32c/src/crc32c_arm64.cc.o -MF contrib/crc32c-cmake/CMakeFiles/_crc32c.dir/__/crc32c/src/crc32c_arm64.cc.o.d -o contrib/crc32c-cmake/CMakeFiles/_crc32c.dir/__/crc32c/src/crc32c_arm64.cc.o -c /root/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc
/root/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc:86:20: error: always_inline function 'vmull_p64' requires target feature 'aes', but would be inlined into function 'ExtendArm64' that is compiled without support for 'aes'
   86 |     t2 = (uint64_t)vmull_p64(crc2, k2);
      |                    ^
/root/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc:87:20: error: always_inline function 'vmull_p64' requires target feature 'aes', but would be inlined into function 'ExtendArm64' that is compiled without support for 'aes'
   87 |     t1 = (uint64_t)vmull_p64(crc1, k1);
      |                    ^
/root/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc:88:20: error: always_inline function 'vmull_p64' requires target feature 'aes', but would be inlined into function 'ExtendArm64' that is compiled without support for 'aes'
   88 |     t0 = (uint64_t)vmull_p64(crc0, k0);
      |                    ^
3 errors generated.
```

--- a/cmake/cpu_features.cmake
+++ b/cmake/cpu_features.cmake
@@ -13,6 +13,14 @@ if (ARCH_NATIVE)
     set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=native")
     list(APPEND RUSTFLAGS_CPU "-C" "target-feature=native")
 
+    macro(GET_CPU_FEATURES TEST_FEATURE_RESULT)
+       execute_process(
+            COMMAND sh -c "clang -E - -march=native -###"
+            INPUT_FILE /dev/null
+            OUTPUT_QUIET
+            ERROR_VARIABLE ${TEST_FEATURE_RESULT})
+    endmacro()
+
     macro(TEST_CPU_FEATURE TEST_FEATURE_RESULT feat flag)
         if (${TEST_FEATURE_RESULT} MATCHES "\"\\+${feat}\"")
             set(${flag} ON)
@@ -21,20 +29,12 @@ if (ARCH_NATIVE)
         endif ()
     endmacro()
 
-    macro(RUN_TEST_CPU_FEATURE TEST_FEATURE_RESULT)
-       execute_process(
-            COMMAND sh -c "clang -E - -march=native -###"
-            INPUT_FILE /dev/null
-            OUTPUT_QUIET
-            ERROR_VARIABLE ${TEST_FEATURE_RESULT})
-    endmacro()
-
     # Populate the ENABLE_ option flags. This is required for the build of some third-party dependencies, specifically snappy, which
     # (somewhat weirdly) expects the relative SNAPPY_HAVE_ preprocessor variables to be populated, in addition to the microarchitecture
     # feature flags being enabled in the compiler. This fixes the ARCH_NATIVE flag by automatically populating the ENABLE_ option flags
     # according to the current CPU's capabilities, detected using clang.
     if (ARCH_AMD64)
-        RUN_TEST_CPU_FEATURE (TEST_FEATURE_RESULT)
+        GET_CPU_FEATURES (TEST_FEATURE_RESULT)
         message(FATAL_ERROR "PMO: ${TEST_FEATURE_RESULT}")
 
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} ssse3 ENABLE_SSSE3)
@@ -50,7 +50,7 @@ if (ARCH_NATIVE)
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} bmi2 ENABLE_BMI2)
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} aes ENABLE_AES)
     elseif (ARCH_AARCH64)
-        RUN_TEST_CPU_FEATURE (TEST_FEATURE_RESULT)
+        GET_CPU_FEATURES (TEST_FEATURE_RESULT)
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} aes ENABLE_AES)
     endif ()
 

--- a/cmake/cpu_features.cmake
+++ b/cmake/cpu_features.cmake
@@ -35,7 +35,6 @@ if (ARCH_NATIVE)
     # according to the current CPU's capabilities, detected using clang.
     if (ARCH_AMD64)
         GET_CPU_FEATURES (TEST_FEATURE_RESULT)
-        message(FATAL_ERROR "PMO: ${TEST_FEATURE_RESULT}")
 
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} ssse3 ENABLE_SSSE3)
         TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} sse4.1 ENABLE_SSE41)

--- a/cmake/cpu_features.cmake
+++ b/cmake/cpu_features.cmake
@@ -13,36 +13,45 @@ if (ARCH_NATIVE)
     set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=native")
     list(APPEND RUSTFLAGS_CPU "-C" "target-feature=native")
 
+    macro(TEST_CPU_FEATURE TEST_FEATURE_RESULT feat flag)
+        if (${TEST_FEATURE_RESULT} MATCHES "\"\\+${feat}\"")
+            set(${flag} ON)
+        else ()
+            set(${flag} OFF)
+        endif ()
+    endmacro()
+
+    macro(RUN_TEST_CPU_FEATURE TEST_FEATURE_RESULT)
+       execute_process(
+            COMMAND sh -c "clang -E - -march=native -###"
+            INPUT_FILE /dev/null
+            OUTPUT_QUIET
+            ERROR_VARIABLE ${TEST_FEATURE_RESULT})
+    endmacro()
+
     # Populate the ENABLE_ option flags. This is required for the build of some third-party dependencies, specifically snappy, which
     # (somewhat weirdly) expects the relative SNAPPY_HAVE_ preprocessor variables to be populated, in addition to the microarchitecture
     # feature flags being enabled in the compiler. This fixes the ARCH_NATIVE flag by automatically populating the ENABLE_ option flags
     # according to the current CPU's capabilities, detected using clang.
     if (ARCH_AMD64)
-        execute_process(
-            COMMAND sh -c "clang -E - -march=native -###"
-            INPUT_FILE /dev/null
-            OUTPUT_QUIET
-            ERROR_VARIABLE TEST_FEATURE_RESULT)
+        RUN_TEST_CPU_FEATURE (TEST_FEATURE_RESULT)
+        message(FATAL_ERROR "PMO: ${TEST_FEATURE_RESULT}")
 
-        macro(TEST_AMD64_FEATURE TEST_FEATURE_RESULT feat flag)
-            if (${TEST_FEATURE_RESULT} MATCHES "\"\\+${feat}\"")
-                set(${flag} ON)
-            else ()
-                set(${flag} OFF)
-            endif ()
-        endmacro()
-
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} ssse3 ENABLE_SSSE3)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} sse4.1 ENABLE_SSE41)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} sse4.2 ENABLE_SSE42)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} vpclmulqdq ENABLE_PCLMULQDQ)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} popcnt ENABLE_POPCNT)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} avx ENABLE_AVX)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} avx2 ENABLE_AVX2)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} avx512f ENABLE_AVX512)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} avx512vbmi ENABLE_AVX512_VBMI)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} bmi ENABLE_BMI)
-        TEST_AMD64_FEATURE (${TEST_FEATURE_RESULT} bmi2 ENABLE_BMI2)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} ssse3 ENABLE_SSSE3)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} sse4.1 ENABLE_SSE41)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} sse4.2 ENABLE_SSE42)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} vpclmulqdq ENABLE_PCLMULQDQ)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} popcnt ENABLE_POPCNT)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} avx ENABLE_AVX)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} avx2 ENABLE_AVX2)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} avx512f ENABLE_AVX512)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} avx512vbmi ENABLE_AVX512_VBMI)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} bmi ENABLE_BMI)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} bmi2 ENABLE_BMI2)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} aes ENABLE_AES)
+    elseif (ARCH_AARCH64)
+        RUN_TEST_CPU_FEATURE (TEST_FEATURE_RESULT)
+        TEST_CPU_FEATURE (${TEST_FEATURE_RESULT} aes ENABLE_AES)
     endif ()
 
 elseif (ARCH_AARCH64)

--- a/contrib/crc32c-cmake/CMakeLists.txt
+++ b/contrib/crc32c-cmake/CMakeLists.txt
@@ -12,10 +12,14 @@ set(CRC32C_SOURCES
     "${CRC32C_DIR}/src/crc32c.cc"
 )
 
-if(ARCH_AARCH64)
+# crc32c_arm64.cc uses the vmull_p64 instruction, which is only available when building
+# with the `aes` flag, which is included among others under the `crypto` flag.
+# We only add the `crypto` flag when building on modern ARM CPUs, so we can't
+# build that file on older ARM CPUs.
+if(ARCH_AARCH64 AND (NOT NO_ARMV81_OR_HIGHER))
   include_directories(${CRC32C_CMAKE_DIR}/aarch64)
   list(APPEND CRC32C_SOURCES "${CRC32C_DIR}/src/crc32c_arm64.cc")
-else()
+elseif(ARCH_AMD64)
   if(ENABLE_SSE42)
     include_directories(${CRC32C_CMAKE_DIR}/x86_64/sse42)
     list(APPEND CRC32C_SOURCES "${CRC32C_DIR}/src/crc32c_sse42.cc")

--- a/contrib/crc32c-cmake/CMakeLists.txt
+++ b/contrib/crc32c-cmake/CMakeLists.txt
@@ -12,11 +12,14 @@ set(CRC32C_SOURCES
     "${CRC32C_DIR}/src/crc32c.cc"
 )
 
-# crc32c_arm64.cc uses the vmull_p64 instruction, which is only available when building
-# with the `aes` flag.
-if(ARCH_AARCH64 AND ENABLE_AES)
+if(ARCH_AARCH64)
   include_directories(${CRC32C_CMAKE_DIR}/aarch64)
-  list(APPEND CRC32C_SOURCES "${CRC32C_DIR}/src/crc32c_arm64.cc")
+
+  # crc32c_arm64.cc uses the vmull_p64 instruction, which is only available when building
+  # with the `aes` flag.
+  if (ENABLE_AES)
+    list(APPEND CRC32C_SOURCES "${CRC32C_DIR}/src/crc32c_arm64.cc")
+  endif()
 elseif(ARCH_AMD64)
   if(ENABLE_SSE42)
     include_directories(${CRC32C_CMAKE_DIR}/x86_64/sse42)

--- a/contrib/crc32c-cmake/CMakeLists.txt
+++ b/contrib/crc32c-cmake/CMakeLists.txt
@@ -13,10 +13,8 @@ set(CRC32C_SOURCES
 )
 
 # crc32c_arm64.cc uses the vmull_p64 instruction, which is only available when building
-# with the `aes` flag, which is included among others under the `crypto` flag.
-# We only add the `crypto` flag when building on modern ARM CPUs, so we can't
-# build that file on older ARM CPUs.
-if(ARCH_AARCH64 AND (NOT NO_ARMV81_OR_HIGHER))
+# with the `aes` flag.
+if(ARCH_AARCH64 AND ENABLE_AES)
   include_directories(${CRC32C_CMAKE_DIR}/aarch64)
   list(APPEND CRC32C_SOURCES "${CRC32C_DIR}/src/crc32c_arm64.cc")
 elseif(ARCH_AMD64)

